web/packages/design/src/utils/useResizeObserver.ts

```diff
@@ -0,0 +1,60 @@
+/**
+ * Teleport
+ * Copyright (C) 2024 Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import { RefObject, useCallback, useLayoutEffect } from 'react';
+
+/**
+ * useResizeObserver sets up a ResizeObserver for ref and calls callback on each resize.
+ *
+ * It does not fire if ref.current.contentRect.height is zero, to account for a special case in
+ * Connect where tabs are hidden using `display: none;`.
+ *
+ * Uses a layout effect underneath.
+ */
+export function useResizeObserver(
+  ref: RefObject<HTMLElement>,
+  callback: () => void,
+  { enabled = true }
+): void {
+  const effect = useCallback(() => {
+    if (!ref.current || !enabled) {
+      return;
+    }
+
+    const observer = new ResizeObserver(entries => {
+      const element = entries[0];
+
+      // In Connect, when a tab becomes active, its outermost DOM element switches from `display:
+      // none` to `display: flex`. This callback is then fired with the height reported as zero.
+      // To avoid unnecessary calls to callback, return early here.
+      if (element.contentRect.height === 0) {
+        return;
+      }
+
+      callback();
+    });
+
+    observer.observe(ref.current);
+
+    return () => {
+      observer.disconnect();
```

Comment:
`ref.current` is not supposed to be used as a dep, as updating what `ref.current` points to does not cause a re-render, making it almost a noop as a dep. `ref` is in the deps so that if the consumer of `useResizeObserver` changes what ref is passed to the hook, the effect is adjusted accordingly.

It's the caller's responsibility to ensure that `callback` is not changing on each render. I know that we often don't adhere to this. But ultimately it's something we need to address instead of adjusting new code just in case that `callback` doesn't have a stable identity. That's also why I didn't refactor existing callsites to use this hook and instead left a TODO comment – I didn't have time to ensure that they all use stable callbacks. I should probably mention that in the TODO comments that the callback should be made stable.

And the reason I made `useResizeObserver` in the first place, which I forgot to mention, is to make sure that any shared code that uses a resize observer accounts for the tabs in Connect being hidden with `display: none`.

Reply:
I know that `ref.current` is not supposed to be in the dependencies. Perhaps I framed my comment in a confusing way. What I meant is that _if_ the ref pointer changes because the element in question gets removed from the DOM and replaced with another one, this hook may misbehave. If it gets called again with the same callback and `enabled` value, it will not install the observer on the newly supplied node even if it changed.

It's certainly a weird corner case, but if we want this to be a reusable hook, instead of a component implementation detail, we might want to look into whether I'm actually right and if it's a problem.

I'm not really sure what would be the best way to solve it, but I guess one potential solution would be not to pass a ref to this hook, but instead _return_ a function from this hook that you can then pass as a `ref` parameter of the rendered node. So that whenever a new node is observed, we can immediately react and install an observer on it.